### PR TITLE
feat: Add an API to fetch the list of core config properties

### PR DESCRIPTION
**Pull request by prateek3255**
## Summary of change
-  Adds an API to fetch the list of core config properties

## Related issues

N/A

## Test Plan

N/A

## Documentation changes
N/A

## Checklist for important updates

- [x] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [x] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Update function `getValidFields` in `io/supertokens/config/CoreConfig.java` if new aliases were added for any core config (similar to the `access_token_signing_key_update_interval` config alias).
- [x] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.
- [ ] If added a foreign key constraint on `app_id_to_user_id` table, make sure to delete from this table when deleting the user as well if `deleteUserIdMappingToo` is false.
## Remaining TODOs for this PR

N/A

### CODE REVIEW

File: src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    protected boolean checkAPIKey(HttpServletRequest req) {
+        return false;
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonArray config = CoreConfig.getConfigFieldsJson();
```
Comment (sattvikc) [inline src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java:50]: We need to include config from the plugin as well

File: src/main/java/io/supertokens/config/annotations/DifferentAcrossTenants.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *    License for the specific language governing permissions and limitations
+ *    under the License.
+ */
+
+ package io.supertokens.config.annotations;
+
+ import java.lang.annotation.ElementType;
```
Comment (sattvikc) [inline src/main/java/io/supertokens/config/annotations/DifferentAcrossTenants.java:19]: We might not need this annotation, instead we can infer - Is JsonProperty and does not have ConfigYamlOnly and NotConflictingInApp annotations. Cross check if this is correct

File: src/main/java/io/supertokens/config/CoreConfig.java [hunk 3]
```diff
@@ -718,16 +767,66 @@ public void createLoggingFile(Main main) throws IOException {
...
+                fieldJson.addProperty("type", type);
+                result.add(fieldJson);
+
+            } catch (NoSuchFieldException e) {
+                continue;
+            }
+        }
+        return result;
+    }
+
```
Comment (sattvikc) [inline src/main/java/io/supertokens/config/CoreConfig.java:829]: in the normalizeAndValidate function in this file, see if you can use the EnumProperty annotation for the validation.

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java (hunk 1 above)
Comment (prateek3255) [reply src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java]: Added this and opened separate PR for Postgres plugin

File: src/main/java/io/supertokens/config/annotations/DifferentAcrossTenants.java (hunk 2 above)
Comment (prateek3255) [reply src/main/java/io/supertokens/config/annotations/DifferentAcrossTenants.java]: Yeah you are right removed this

File: src/main/java/io/supertokens/config/CoreConfig.java (hunk 3 above)
Comment (prateek3255) [reply src/main/java/io/supertokens/config/CoreConfig.java]: Yeah added checks for that in the method

File: src/main/java/io/supertokens/config/CoreConfig.java [hunk 4]
```diff
@@ -46,117 +50,144 @@
...
     private int core_config_version = -1;
 
     @NotConflictingInApp
     @JsonProperty
+    @ConfigDescription("Time in seconds for how long an access token is valid for. [Default: 3600 (1 hour)]")
     private long access_token_validity = 3600; // in seconds
 
     @NotConflictingInApp
     @JsonProperty
+    @ConfigDescription("Deprecated, please see changelog. Only used in CDI<=2.18\n If true, allows for immediate revocation of any access token. Keep in mind that setting this to true will result in a db query for each API call that requires authentication. (Default: false)")
```
Comment (sattvikc) [inline src/main/java/io/supertokens/config/CoreConfig.java:70]: `\n` not necessary

File: src/test/java/io/supertokens/test/CoreConfigListAPITest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+ import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+ import io.supertokens.ProcessState;
+ import io.supertokens.httpRequest.HttpRequest;
+ import org.junit.AfterClass;
+ import org.junit.Before;
+ import org.junit.Rule;
+ import org.junit.Test;
+ import org.junit.rules.TestRule;
```
Comment (sattvikc) [inline src/test/java/io/supertokens/test/CoreConfigListAPITest.java:27]: there seem to be some indentation issue

File: src/test/java/io/supertokens/test/CoreConfigListAPITest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            String type = config.get("type").getAsString();
+            assertTrue(type.equals("number") || type.equals("boolean") || type.equals("string") || type.equals("enum"));
+        }
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+     }
+ 
+ }
+
```
Comment (sattvikc) [inline src/test/java/io/supertokens/test/CoreConfigListAPITest.java:74]: Add test for protected fields

File: src/test/java/io/supertokens/test/CoreConfigListAPITest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            assertTrue(config.get("type").getAsJsonPrimitive().isString());
+            String type = config.get("type").getAsString();
+            assertTrue(type.equals("number") || type.equals("boolean") || type.equals("string") || type.equals("enum"));
+        }
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+     }
+ 
+ }
```
Comment (sattvikc) [inline src/test/java/io/supertokens/test/CoreConfigListAPITest.java:73]: Also can you add a test that can match the description between `config.yaml` file and the description annotations ? This is to ensure they are correct always. Also try to ensure all the properties are present in the config.yaml and devConfig.yaml file.

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/test/java/io/supertokens/test/CoreConfigListAPITest.java (hunk 5 above)
Comment (prateek3255) [reply src/test/java/io/supertokens/test/CoreConfigListAPITest.java]: Fixed

File: src/test/java/io/supertokens/test/CoreConfigListAPITest.java (hunk 6 above)
Comment (prateek3255) [reply src/test/java/io/supertokens/test/CoreConfigListAPITest.java]: Added in the same test to ensure that protected fields are not there in the API result

File: src/main/java/io/supertokens/config/CoreConfig.java (hunk 3 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/config/CoreConfig.java:853]: I don't see this change

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/config/CoreConfig.java [hunk 8]
```diff
@@ -590,6 +631,28 @@ void normalizeAndValidate(Main main, boolean includeConfigFilePath) throws Inval
             }
         }
 
+        try {
+            String[] allowedPasswordHashingAlgos = CoreConfig.class.getDeclaredField("password_hashing_alg")
```
Comment (sattvikc) [inline src/main/java/io/supertokens/config/CoreConfig.java:635]: you can make the generic, iterate thru all properties and check for this annotation. we might miss adding this logic if we add new enum properties.

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/config/CoreConfig.java (hunk 8 above)
Comment (prateek3255) [reply src/main/java/io/supertokens/config/CoreConfig.java]: Done! Although I wasn't able to figure out a way to test it. Can you check if it looks good or let me know if you know to check that it works correctly?

File: src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        ArrayList<ConfigFieldInfo> config = CoreConfig.getConfigFieldsInfo(main);
+        JsonObject result = new JsonObject();
+
+        JsonArray configJson = new JsonArray();
+        for (ConfigFieldInfo field : config) {
+            JsonObject fieldJson = new JsonObject();
```
Comment (sattvikc) [inline src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java:61]: You could potentially JsonEncode the array directly, by adding some json annotation in the class

File: src/main/java/io/supertokens/config/CoreConfig.java [hunk 10]
```diff
@@ -718,14 +782,70 @@ public void createLoggingFile(Main main) throws IOException {
...
+            } catch (NoSuchFieldException e) {
+                continue;
             }
         }
+
+        ArrayList<ConfigFieldInfo> storageFields = StorageLayer.getBaseStorage(main).getConfigFieldsInfo();
+
+        result.addAll(storageFields);
+
+        return result;
```
Comment (sattvikc) [inline src/main/java/io/supertokens/config/CoreConfig.java:848]: Move this to the API layer itself and remove from here.

**Review by sattvikc — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java (hunk 9 above)
Comment (prateek3255) [reply src/main/java/io/supertokens/webserver/api/core/CoreConfigListAPI.java]: Used a different way

File: src/main/java/io/supertokens/config/CoreConfig.java (hunk 10 above)
Comment (prateek3255) [reply src/main/java/io/supertokens/config/CoreConfig.java]: Moved it to API layer

File: src/test/java/io/supertokens/test/CoreConfigListAPITest.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                    propertyDescription = propertyDescription.trim();
+
+                    propertyDescriptions.put(propertyKey, propertyDescription);
+                }
+            }
+
+            for (String fieldId : CoreConfig.getValidFields()) {
+                if (Arrays.asList(ignoreProperties).contains(fieldId)) {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (prateek3255) [reply src/test/java/io/supertokens/test/CoreConfigListAPITest.java:229]: Done

**Pull request merged**